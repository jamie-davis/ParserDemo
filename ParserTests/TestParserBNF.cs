using Xunit;
using TestConsoleLib;
using TestConsoleLib.Testing;
using TestConsole.OutputFormatting;
using System.Linq;
using Parser;

namespace ParserTests
{
    public class TestParserBNF
    {
        [Fact]
        public void InputIsConverted()
        {
            //Arrange
            var testStrings = new [] {
                "1",
                "32-5",
                "(3 *5)+  49",
                "1 + 2 + 3",
                "145 * ((63/2 + 5) - 16)",
                "2 * 2 * (2 * 2)",
                "1 + +",
                "(1 + 5",
                "1 2",
                "+ 1",
                "((",
                ")",
                "+",
                "",
                "(",
                "145 * ((63/2 + 5) - 16))",
                "14.5 * ((6.3/2.4 + 5.5) - 1.6)",
            };

            //Act
            var results = testStrings
                .Select(s => new { String = s, Result = ArithmeticParserBNF.Parse(s)})
                .ToList();

            //Assert
            var output = new Output();
            var report = results.AsReport(rep => rep
                .AddColumn(r => r.String, cc => cc.Heading("Input"))
                .AddColumn(r => r.Result.IsValid, cc => { })
                .AddColumn(r => r.Result.Describe(), cc => cc.Heading("Result"))
                .AddColumn(r => r.Result.ErrorMessage, cc => {})
            );

            output.FormatTable(report);
            output.Report.Verify();
        }
    }
}