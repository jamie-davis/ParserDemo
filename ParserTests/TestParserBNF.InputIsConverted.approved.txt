                          Is                                                                   
Input                     Valid Result                        Error Message                    
------------------------- ----- ----------------------------- ---------------------------------
1                         True  1                                                              
32-5                      True  -[32, 5]                                                       
(3 *5)+  49               True  +[(*[3, 5]), 49]                                               
1 + 2 + 3                 True  +[1, +[2, 3]]                                                  
145 * ((63/2 + 5) - 16)   True  *[145, (-[(+[/[63, 2], 5]),                                    
                                16])]                                                          
2 * 2 * (2 * 2)           True  *[2, *[2, (*[2, 2])]]                                          
1 + +                     False                               Unable to interpret calculation  
                                                              at "+ +"                         
(1 + 5                    False                               Unable to interpret calculation  
                                                              at "( 1 + 5"                     
1 2                       False                               Unable to interpret calculation  
                                                              at "2"                           
+ 1                       False                               Unable to interpret calculation  
                                                              at "+ 1"                         
((                        False                               Unable to interpret calculation  
                                                              at "( ("                         
)                         False                               Unable to interpret calculation  
                                                              at ")"                           
+                         False                               Unable to interpret calculation  
                                                              at "+"                           
                          False                               Unable to interpret calculation  
                                                              at ""                            
(                         False                               Unable to interpret calculation  
                                                              at "("                           
145 * ((63/2 + 5) - 16))  False                               Unable to interpret calculation  
                                                              at ")"                           
14.5 * ((6.3/2.4 + 5.5) - True  *[14.5, (-[(+[/[6.3, 2.4], 5.                                  
1.6)                            5]), 1.6])]                                                    
                                                                                               